source/common/router/router.cc

```diff
@@ -602,6 +602,8 @@ void Filter::onUpstreamHeaders(const uint64_t response_code, Http::HeaderMapPtr&
     headers->addReferenceKey(Http::Headers::get().SetCookie, header_value);
   }
 
+  route_entry_->finalizeResponseHeaders(*headers);
```

Comment:
I went back and forth on this when we originally started expanding this out a while back (I think @rshriram did that work), but I'm not sure if we should do this here vs. inside the finalize response headers utility in the connection manager. The argument for putting here is consistency, the argument against is that theoretically someone could want to add response headers w/o using the router filter. I guess I'm inclined to move it here as you have done and just document that the router filter is the thing that does this. Any competing thoughts?

Reply:
It seemed more consistent to me to put it here. Seems like the same thing applies to request headers, although perhaps they're less likely to be of use outside the router filter.

For access outside the router filter, is it sufficient to add `finalizeRequestHeaders` and `finalizeResponseHeaders` to `Router::VirtualHost` and `Router::Config`?

Reply:
I think for consistency I would just do what you did. I would add a TODO to figure out if this is ever needed outside of router context. If so, we can fix it then?

Reply:
Sounds good.